File: src/services/aiSettings.ts
import type { Env, Viewer } from "../types";
import { HttpError, json, readJson, safeJsonParse } from "../utils";

const AI_SETTINGS_KEY = "ai.settings";
const DEFAULT_AI_BASE_URL = "https://api.openai.com/v1";
const DEFAULT_AI_MODEL = "gpt-4o-mini";

export interface AiSettings {
  baseUrl: string;
  model: string;
  apiKey: string;
}

export interface AiSettingsClient {
  baseUrl: string;
  model: string;
  configured: boolean;
}

export function sanitizeAiSettingsForClient(settings: AiSettings): AiSettingsClient {
  return {
    baseUrl: settings.baseUrl,
    model: settings.model,
    configured: Boolean(settings.apiKey.trim()),
  };
}

export function mergeAiSettingsUpdate(previous: AiSettings, update: Partial<AiSettings>): AiSettings {
  return {
    baseUrl: normalizeBaseUrl(update.baseUrl ?? previous.baseUrl),
    model: String(update.model ?? previous.model).trim() || previous.model || DEFAULT_AI_MODEL,
    apiKey: update.apiKey === undefined || !String(update.apiKey).trim()
      ? previous.apiKey
      : String(update.apiKey).trim(),
  };
}

export async function resolveAiRuntimeSettings(env: Env): Promise<AiSettings> {
  const stored = await readStoredAiSettings(env);
  return {
    baseUrl: normalizeBaseUrl(stored.baseUrl || env.AI_BASE_URL || DEFAULT_AI_BASE_URL),
    model: stored.model || env.AI_MODEL || DEFAULT_AI_MODEL,
    apiKey: stored.apiKey || env.AI_API_KEY || "",
  };
}

export async function getAiSettings(env: Env, viewer: Viewer): Promise<Response> {
  if (viewer.role !== "ADMIN") return json({ error: "Forbidden" }, 403);
  return json({ settings: sanitizeAiSettingsForClient(await resolveAiRuntimeSettings(env)) });
}

export async function updateAiSettings(request: Request, env: Env, viewer: Viewer): Promise<Response> {
  if (viewer.role !== "ADMIN") return json({ error: "Forbidden" }, 403);
  const body = await readJson<Partial<AiSettings>>(request);
  const previous = await readStoredAiSettings(env);
  const next = mergeAiSettingsUpdate(previous, body);
  await writeStoredAiSettings(env, next);
  return json({ settings: sanitizeAiSettingsForClient(next) });
}

export async function testAiSettings(request: Request, env: Env, viewer: Viewer): Promise<Response> {
  if (viewer.role !== "ADMIN") return json({ error: "Forbidden" }, 403);
  const body = await readJson<Partial<AiSettings>>(request);
  const base = await resolveAiRuntimeSettings(env);
  const settings = mergeAiSettingsUpdate(base, body);
  if (!settings.apiKey) return json({ error: "AI API Key is required" }, 400);

  const response = await fetch(`${settings.baseUrl.replace(/\/+$/, "")}/chat/completions`, {
    method: "POST",
    headers: {
      "Authorization": `Bearer ${settings.apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({
      model: settings.model,
      temperature: 0,
      messages: [
        { role: "system", content: "Return ok." },
        { role: "user", content: "ping" },
      ],
      max_tokens: 8,
    }),
  });
  if (!response.ok) throw new HttpError(`AI API returned HTTP ${response.status}`, 502);
  return json({ ok: true });
}

async function readStoredAiSettings(env: Env): Promise<AiSettings> {
  const row = await env.DB.prepare("SELECT value FROM system_setting WHERE name = ?")
    .bind(AI_SETTINGS_KEY)
    .first<{ value: string }>();
  const stored = safeJsonParse<Partial<AiSettings>>(row?.value ?? "{}", {});
  return {
    baseUrl: normalizeBaseUrl(stored.baseUrl || DEFAULT_AI_BASE_URL),
    model: String(stored.model || DEFAULT_AI_MODEL).trim() || DEFAULT_AI_MODEL,
    apiKey: String(stored.apiKey || "").trim(),
  };
}

async function writeStoredAiSettings(env: Env, settings: AiSettings): Promise<void> {
  await env.DB.prepare(`
    INSERT INTO system_setting (name, value, description)
    VALUES (?, ?, ?)
    ON CONFLICT(name) DO UPDATE SET value = excluded.value
  `).bind(AI_SETTINGS_KEY, JSON.stringify(settings), "AI model settings").run();
}

function normalizeBaseUrl(value: string): string {
  const raw = String(value || DEFAULT_AI_BASE_URL).trim() || DEFAULT_AI_BASE_URL;
  let url: URL;
  try {
    url = new URL(raw);
  } catch {
    throw new HttpError("Invalid AI Base URL", 400);
  }
  if (url.protocol !== "http:" && url.protocol !== "https:") {
    throw new HttpError("Invalid AI Base URL", 400);
  }
  url.hash = "";
  url.search = "";
  return url.toString().replace(/\/+$/, "");
}

File: src/ai_settings.rs
use super::*;

pub(crate) async fn get_ai_settings(
    env: &Env,
    viewer: &Viewer,
) -> std::result::Result<Response, AppError> {
    require_admin(viewer)?;
    let settings = resolve_ai_settings(env).await?;
    json_response(json!({ "settings": public_ai_settings(&settings) }), 200).map_err(AppError::from)
}

pub(crate) async fn update_ai_settings(
    req: &mut Request,
    env: &Env,
    viewer: &Viewer,
) -> std::result::Result<Response, AppError> {
    require_admin(viewer)?;
    let body: Value = req
        .json()
        .await
        .map_err(|_| AppError::new(400, "Invalid JSON"))?;
    let previous = read_stored_ai_settings(env).await?;
    let next = merge_ai_settings(&previous, &body)?;
    db(env)?.prepare("INSERT INTO system_setting (name, value, description) VALUES ('ai.settings', ?, 'AI model settings') ON CONFLICT(name) DO UPDATE SET value = excluded.value")
        .bind(&[serde_json::to_string(&next).map_err(|error| AppError::new(500, error.to_string()))?.into()])?
        .run()
        .await?;
    json_response(json!({ "settings": public_ai_settings(&next) }), 200).map_err(AppError::from)
}

pub(crate) async fn test_ai_settings(
    req: &mut Request,
    env: &Env,
    viewer: &Viewer,
) -> std::result::Result<Response, AppError> {
    require_admin(viewer)?;
    let body: Value = req.json().await.unwrap_or_else(|_| json!({}));
    let base = resolve_ai_settings(env).await?;
    let settings = merge_ai_settings(&base, &body)?;
    if settings.api_key.trim().is_empty() {
        return Err(AppError::new(400, "AI API Key is required"));
    }
    let headers = Headers::new();
    headers.set("Authorization", &format!("Bearer {}", settings.api_key))?;
    headers.set("Content-Type", "application/json")?;
    let payload = json!({
        "model": settings.model,
        "temperature": 0,
        "messages": [
            { "role": "system", "content": "Return ok." },
            { "role": "user", "content": "ping" }
        ],
        "max_tokens": 8
    });
    let mut init = RequestInit::new();
    init.with_method(Method::Post)
        .with_headers(headers)
        .with_body(Some(JsValue::from_str(&payload.to_string())));
    let request = Request::new_with_init(
        &format!(
            "{}/chat/completions",
            settings.base_url.trim_end_matches('/')
        ),
        &init,
    )?;
    let response = Fetch::Request(request).send().await?;
    if response.status_code() < 200 || response.status_code() >= 300 {
        return Err(AppError::new(
            502,
            format!("AI API returned HTTP {}", response.status_code()),
        ));
    }
    json_response(json!({ "ok": true }), 200).map_err(AppError::from)
}

pub(crate) async fn resolve_ai_settings(env: &Env) -> std::result::Result<AiSettings, AppError> {
    let stored = read_stored_ai_settings(env).await?;
    Ok(AiSettings {
        base_url: normalize_http_url(
            if stored.base_url.is_empty() {
                env.var("AI_BASE_URL")
                    .map(|value| value.to_string())
                    .unwrap_or_else(|_| "https://api.openai.com/v1".to_string())
            } else {
                stored.base_url
            },
            "Invalid AI Base URL",
        )?,
        model: if stored.model.is_empty() {
            env.var("AI_MODEL")
                .map(|value| value.to_string())
                .unwrap_or_else(|_| "gpt-4o-mini".to_string())
        } else {
            stored.model
        },
        api_key: if stored.api_key.is_empty() {
            env.secret("AI_API_KEY")
                .map(|value| value.to_string())
                .unwrap_or_default()
        } else {
            stored.api_key
        },
    })
}

pub(crate) async fn read_stored_ai_settings(
    env: &Env,
) -> std::result::Result<AiSettings, AppError> {
    let value: Option<String> = db(env)?
        .prepare("SELECT value FROM system_setting WHERE name = 'ai.settings'")
        .first(Some("value"))
        .await?;
    let stored = value.and_then(|text| serde_json::from_str::<AiSettings>(&text).ok());
    Ok(stored.unwrap_or(AiSettings {
        base_url: "https://api.openai.com/v1".to_string(),
        model: "gpt-4o-mini".to_string(),
        api_key: String::new(),
    }))
}

pub(crate) fn merge_ai_settings(
    previous: &AiSettings,
    update: &Value,
) -> std::result::Result<AiSettings, AppError> {
    let base_url = update
        .get("baseUrl")
        .and_then(Value::as_str)
        .unwrap_or(&previous.base_url);
    let model = update
        .get("model")
        .and_then(Value::as_str)
        .unwrap_or(&previous.model)
        .trim();
    let api_key = update
        .get("apiKey")
        .and_then(Value::as_str)
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .unwrap_or(&previous.api_key);
    Ok(AiSettings {
        base_url: normalize_http_url(base_url, "Invalid AI Base URL")?,
        model: if model.is_empty() {
            "gpt-4o-mini".to_string()
        } else {
            model.to_string()
        },
        api_key: api_key.to_string(),
    })
}

pub(crate) fn public_ai_settings(settings: &AiSettings) -> Value {
    json!({
        "baseUrl": settings.base_url,
        "model": settings.model,
        "configured": !settings.api_key.trim().is_empty()
    })
}
